File: raspberryPI3/ros2_ws/src/adas/src/collision_avoidance_node.cpp
#include <cstdio>

#include "interfaces/msg/motors_order.hpp"
#include "interfaces/msg/ultrasonic.hpp"
#include "std_srvs/srv/empty.hpp"

#include "rclcpp/rclcpp.hpp"
#include <chrono>
#include <functional>
#include <memory>

using std::placeholders::_1;

class collision_avoidance : public rclcpp::Node {
public:
  collision_avoidance()
  : Node("collision_avoidance_node")
  {
    subscription_us_data_ = this->create_subscription<interfaces::msg::Ultrasonic>(
    "us_data", 10, std::bind(&collision_avoidance::detectCollisionCallback, this, _1));

    subscription_motors_order_raw_ = this->create_subscription<interfaces::msg::MotorsOrder>(
    "motors_order_raw", 10, std::bind(&collision_avoidance::motorsOrderCallback, this, _1));

    publisher_motors_order_ = this->create_publisher<interfaces::msg::MotorsOrder>("motors_order", 10);

    RCLCPP_INFO(this->get_logger(), "collision_avoidance_node READY");
  }

private:

    struct ultra_data
    {
      int front_left;
      int front_right;
      int front_center;
      int rear_left;
      int rear_right;
      int rear_center;
    };

    struct motors_order
    {
      int right_rear_pwm;
      int left_rear_pwm;
    };

    struct motors_order order;
    struct ultra_data ultras;

    int detectCollisionCallback(const interfaces::msg::Ultrasonic & usData) {

      ultras.front_left = usData.front_left;
      ultras.front_right = usData.front_right;
      ultras.front_center = usData.front_center;

      return detectCollision(ultras);
    }

    int motorsOrderCallback(const interfaces::msg::MotorsOrder & motorsOrder) {

      order.right_rear_pwm = motorsOrder.right_rear_pwm;
      order.left_rear_pwm = motorsOrder.left_rear_pwm;

      return 0;
    }

    int detectCollision(struct ultra_data ultras){
      if (order.right_rear_pwm > 50 || order.left_rear_pwm > 50){
        if (ultras.front_left < 20 || ultras.front_right < 20 || ultras.front_center < 20){
          order.left_rear_pwm = 50;
          order.right_rear_pwm = 50;
          RCLCPP_INFO(this->get_logger(), "Detecting obtacle <20 cm away : Stopping car");
        }
      }

      return 0;
    }

    //Subscribers
    rclcpp::Subscription<interfaces::msg::Ultrasonic>::SharedPtr subscription_us_data_;
    rclcpp::Subscription<interfaces::msg::MotorsOrder>::SharedPtr subscription_motors_order_raw_;

    rclcpp::Publisher<interfaces::msg::MotorsOrder>::SharedPtr publisher_motors_order_;
   
};



int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<collision_avoidance>();
  rclcpp::spin(node);

  rclcpp::shutdown();
  return 0;
}

File: raspberryPI3/ros2_ws/src/adas/adas/collision_avoidance_node.py
import rclpy
from rclpy.node import Node

from interfaces.msg import Ultrasonic, MotorsOrder, MotorsOrderAdas

from rcl_interfaces.msg import SetParametersResult

STOP = 50

STATE_STOP = "state_stop"
STATE_STOP_REAR = "state_stop_rear"
STATE_NOTHING = "state_nothing"
STATE_SLOW = "state_slow"

REFRESH_PERIOD = 0.002

class collision_avoidance(Node):

    def __init__(self):

        super().__init__('collision_avoidance_node')
        self.publisher_motors_order = self.create_publisher(MotorsOrderAdas, 'motors_order_collision', 10)

        self.subscription = self.create_subscription(Ultrasonic,'us_data', self.ultrasonic_callback, 10)
        self.subscription = self.create_subscription(MotorsOrder,'motors_order_raw', self.motors_order_raw_callback, 10)
        self.subscription  # prevent unused variable warning

        self.ultra_front_left = 300
        self.ultra_front_right = 300
        self.ultra_front_center = 300

        self.ultra_rear_left = 300
        self.ultra_rear_right = 300
        self.ultra_rear_center = 300

        self.motor_right_rear_pwm = 0
        self.motor_left_rear_pwm = 0

        self.emergency_stop = False

        self.state = STATE_NOTHING
        self.prev_state = STATE_NOTHING

        self.active = False

        self.declare_parameter("threshold_stop", 20)
        self.declare_parameter("threshold_slow", 40)
        self.declare_parameter("threshold_rear", 30)

        self.declare_parameter("slow_speed_percentage", 30)

        self.threshold_stop = self.get_parameter("threshold_stop").value
        self.threshold_slow = self.get_parameter("threshold_slow").value
        self.threshold_rear = self.get_parameter("threshold_rear").value

        self.slow_speed_percentage = self.get_parameter("slow_speed_percentage").value

        self.add_on_set_parameters_callback(self.param_callback)

        self.timer = self.create_timer(REFRESH_PERIOD, self.detect_collision)

        self.get_logger().info("collision_avoidance_node READY")

    # Parameters callback
    def param_callback(self, params):

        for p in params:

            if (p.name == "threshold_slow"):

                self.threshold_slow = p.value
                self.get_logger().info(f"Parameter {p.name} changed to {self.threshold_slow}")

            if (p.name == "threshold_stop"):

                self.threshold_stop = p.value
                self.get_logger().info(f"Parameter {p.name} changed to {self.threshold_stop}")

            if (p.name == "slow_speed_percentage"):

                self.slow_speed_percentage = p.value
                self.get_logger().info(f"Parameter {p.name} changed to {self.slow_speed_percentage}")

            if (p.name == "threshold_rear"):

                self.threshold_rear = p.value
                self.get_logger().info(f"Parameter {p.name} changed to {self.threshold_rear}")

        return SetParametersResult(successful=True)

    def ultrasonic_callback(self, us_data : Ultrasonic):

        self.ultra_front_left = us_data.front_left
        self.ultra_front_right = us_data.front_right
        self.ultra_front_center = us_data.front_center

        self.ultra_rear_left = us_data.rear_left
        self.ultra_rear_right = us_data.rear_right
        self.ultra_rear_center = us_data.rear_center

    def motors_order_raw_callback(self, msg):

        self.motor_right_rear_pwm = msg.right_rear_pwm
        self.motor_left_rear_pwm = msg.left_rear_pwm

    def detect_collision(self):

        self.active = False
        self.state = STATE_NOTHING
        self.emergency_stop = False
        max_pwm = 50
    
        if ((self.motor_right_rear_pwm >= STOP 
             or self.motor_left_rear_pwm >= STOP) and 
            (self.ultra_front_left < self.threshold_stop or 
            self.ultra_front_right < self.threshold_stop or 
            self.ultra_front_center < self.threshold_stop)):

            self.emergency_stop = True
            self.active = True

            self.state = STATE_STOP

            if self.state != self.prev_state and self.active:
                self.get_logger().info(f"Detecting obstacle <{self.threshold_stop}cm ahead of car : Stopping car")


        elif ((self.motor_right_rear_pwm >= STOP 
             or self.motor_left_rear_pwm >= STOP) and 
             (self.threshold_stop < self.ultra_front_center < self.threshold_slow)):
        
            max_pwm = self.slow_speed_percentage/2
            self.active = True
            
            self.state = STATE_SLOW

            if self.state != self.prev_state and self.active:
                self.get_logger().info(f"Detecting obstacle <{self.threshold_slow}cm : Speed limit {self.slow_speed_percentage}%")

        elif ((self.motor_right_rear_pwm < STOP 
             or self.motor_left_rear_pwm < STOP) and 
            (self.ultra_rear_left < self.threshold_rear or 
            self.ultra_rear_right < self.threshold_rear or 
            self.ultra_rear_center < self.threshold_rear)):
                
            self.emergency_stop = True
            self.active = True
            
            self.state = STATE_STOP_REAR

            if self.state != self.prev_state and self.active:
                self.get_logger().info(f"Detecting obstacle <{self.threshold_rear}cm behind car : Stopping car")

        
        # Publishing
        if self.active or (self.state != self.prev_state):

            msg = MotorsOrderAdas()

            msg.max_pwm = int(max_pwm)
            msg.emergency_stop = self.emergency_stop
            msg.state = self.state
            msg.active = self.active

            self.publisher_motors_order.publish(msg)

        self.prev_state = self.state


def main(args=None):
    rclpy.init(args=args)

    collision_avoidance_node = collision_avoidance()

    rclpy.spin(collision_avoidance_node)

    # Destroy the node explicitly
    # (optional - otherwise it will be done automatically
    # when the garbage collector destroys the node object)
    collision_avoidance_node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()